File: 프로그래머스/LV2/코딩테스트 연습/깊이,너비 우선 탐색(DFS,BFS)/전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <iostream>

using namespace std;

void go(int tower, vector<vector<int>>& wires, set<int>& visit, vector<int> first_wire)
{
    if(visit.find(tower) != visit.end())return;
    visit.insert(tower);

    for(vector<int> wire : wires)
    {
        if(wire == first_wire)continue;
        auto it = find(wire.begin(), wire.end(), tower);
        if(it != wire.end())
        {
            int next = it - wire.begin() ? wire[0] : wire[1];
            go(next, wires, visit, first_wire);
        }
    }
}

int solution(int n, vector<vector<int>> wires) 
{
    int ans = 101;
    for(vector<int> wire : wires)
    {
        set<int> left, right;
        go(wire[0], wires, left, wire);
        go(wire[1], wires, right, wire);
        ans = min(ans, abs((int)left.size() - (int)right.size()));
    }
    return ans;
}

File: 프로그래머스/LV2/코딩테스트 연습/깊이,너비 우선 탐색(DFS,BFS)/전력망을 둘로 나누기/전력망을 둘로 나누기.py
def go(tower, wires, visit, first_wire):
    if tower in visit:
        return
    visit.add(tower)

    for wire in wires:
        if wire == first_wire:
            continue
        if tower in wire:
            next = wire[0] if wire.index(tower) == 1 else wire[1]
            go(next, wires, visit, first_wire)

def solution(n, wires):
    ans = 101

    for wire in wires:
        left, right = set(), set()
        go(wire[0], wires, left, wire)
        go(wire[1], wires, right, wire)
        ans = min(ans, abs(len(left)-len(right)))
    return ans